### DISCUSSION TITLE
How can I ignore all tables except those in a predefined list?

### DISCUSSION THREAD

**Discussion by AaronBeaudoin** [directus/directus#11001] [Feature Requests]
Hello,

I am experimenting with using Directus alongside an existing [Odoo](https://www.odoo.com) database. If you're not familiar with Odoo, it's an open-source ERP system which uses PostgreSQL as a database backend.

The database already has hundreds of tables and I don't plan to use 99% of them via Directus.

What I would like is to have my tables continue to be created and managed via Odoo's ORM and then select a very **limited subset** of them to be accessible by Directus. I would like Directus to **ignore all other tables** which I have not explicitly told it to look at.

I noticed there is a `DB_EXCLUDE_TABLES` environment variable that can be set, but what I'm looking for is essentially the opposite of this. I want Directus by default to not include any tables, and only look at tables in a list I pass to it.

The magic of this for me would be that I can have Odoo manage the tables so that they are still fully integrated into Odoo's ORM and environment, but have Directus act as a user-friendly UI for managing certain tables and as an API access point to allow external applications to access those tables as well.

How do I achieve this? Specifically, how do I have Directus only look at tables in a predefined list I manually give it?

**EDIT:** _Additional info on why this is such a problem: In my Directus interface which I've spun up for testing, all of the hundreds of database tables show up for me not only in the "Data Model" settings page, which is frustrating because I don't need to see any of them in Directus, but they also show up under Roles and Permissions and in the "Content" module. In all 3 places they add an insane amount of unnecessary scrolling. `DB_EXCLUDE_TABLES` is not a solution here because as tables are added and removed via the Odoo ORM I would need to constantly maintain a full list of them in order to keep the environment variable up-to-date._

**Comment by oreilles** [comment]
While this isn't possible at the moment, an alternative would be to create a Postgres user that can only access the tables you want to, then use that user in the connection to the database.

A good implementation that would fulfill your request would be to interpret the DB_EXCLUDE_TABLES as a list of Regexp instead of a list of strings. You would then be able to set `DB_EXCLUDE_TABLES=/^(?!some_table|some_other_table$).*/`. This will most likely have other useful application. @rijkvanzanten what do you think? 🙂

**Reply by AaronBeaudoin** [reply]
Using a Postgres user is an idea I hadn't thought of. Wouldn't that user need additional permissions to create and modify tables so Directus can work with the `directus_<whatever>` tables though? Assuming Directus has never been used in a given database, what would be the minimum user permissions necessary for Directus to be able to create it's tables but also have access limited to a predefined list of other tables?

**Reply by oreilles** [reply]
That's a good point, I assumed that your database was already populated. Indeed that would be an issue - there would also be issues in cases where migrations would try to create new system tables, which does happen every so often. You would then need to connect to the database with a another unrestricted user to bootstrap/migrate the database... not so practical in fine.

**Reply by benhaynes** [reply]
We had this in previous versions by having an "allow list" or "deny list" in the ENV or settings could be an option. This could also use a regex or wildcard to be more open-ended. Just a thought. @rijkvanzanten Do you remember how we did it before?

**Reply by rijkvanzanten** [reply]
@benhaynes iirc it was always a denylist, not an allow list 🤔 similar as to how we're doing it now

**Reply by benhaynes** [reply]
Yeah, that's what I remember. Possible to make that a RegEx (or something else) so it can be more dynamic?

**Reply by rijkvanzanten** [reply]
Yeah, but that still won't allow you to make it an allowlist rather than a denylist 🤔

**Reply by benhaynes** [reply]
True! But you could remove temporary tables (etc) if they're prefixed or have a selectable naming system.

**Reply by AaronBeaudoin** [reply]
If I might chime again, this might seem like a small feature, but for me it singlehandedly opens up Directus to being viable in my use case. Not only does it open up the possibility of using Directus sanely and practically with an existing database which is already controlled by another application, it's also helpful for making transitioning over to Directus **incrementally** a possibility.

**Reply by AaronBeaudoin** [reply]
In the meantime while the feature is unavailable, is there a way for me to completely hide certain tables from the Directus UI entirely?

**Comment by AaronBeaudoin** [comment]
Hey, just wondering if you guys have any plans to add this feature at any point in the future. I've been holding off moving forward with using Directus for my team until I can avoid blasting front-end users with hundreds of tables that shouldn't show up there.

**Reply by joselcvarela** [reply]
You can try to make a PR for it. Maybe add `DB_INCLUDE_TABLES` env var and do the reverse logic here:
https://github.com/directus/directus/blob/main/api/src/services/collections.ts#L253-L255

**Reply by AaronBeaudoin** [reply]
Done! #11828

**Reply by rijkvanzanten** [reply]
> You can try to make a PR for it. Maybe add `DB_INCLUDE_TABLES` env var and do the reverse logic here:
> https://github.com/directus/directus/blob/main/api/src/services/collections.ts#L253-L255

The problem with that is that we then have two environment variables controlling the same thing, where one will directly affect the settings of the other. I think that's a little too confusing. I'd like to find a solution where you configure the allow/deny in a single configuration set. Either through a way to add wildcards and negation into _exclude_ (everything, except: a/b/c) or by adding a different env var that "flips" between allow/deny, like `DB_TABLES=a,b,c` `DB_TABLES_ACTION=Allow` or something along those lines

**Reply by AaronBeaudoin** [reply]
For anyone coming here, this discussion has largely moved over to #11828. Check over there to continue following.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
